sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala

```diff
@@ -208,20 +208,33 @@ object SubExprUtils extends PredicateHelper {
    */
   def getOuterReferences(expr: Expression): Seq[Expression] = {
     val outerExpressions = ArrayBuffer.empty[Expression]
-    def collectOutRefs(input: Expression): Unit = input match {
+
+    def collectOutRefs(input: Expression): Boolean = input match {
       case a: AggregateExpression if containsOuter(a) =>
         if (a.references.nonEmpty) {
           throw QueryCompilationErrors.mixedRefsInAggFunc(a.sql, a.origin)
         } else {
           // Collect and update the sub-tree so that outer references inside this aggregate
           // expression will not be collected. For example: min(outer(a)) -> min(a).
-          val newExpr = stripOuterReference(a)
-          outerExpressions += newExpr
+          // delay collecting outer expression as we want to go as much up as possible
```

Comment:
I wonder if it would make sense to handle like `sum(a) + 1` like outer expressions too?

Reply:
I thought about that, but there are situations which I am not sure if they are correct or meaningful, related to binary operators.
for eg projection as (sum(a) + sum(b) ) .. 
Not sure if this is a valid expression involving aggregates in project and having clause.

Reply:
But you are right.. I suppose so long as any expression is such that  so long as only one of the child returns "true", then the current expression can be part of the outer ref.
so basically it means that as long as the situation is not like.  (sum(a) + sum(b)  +  8, types... the expression can be considered to be referencable using outer ref.
Will make that change..

Reply:
implemented..

Reply:
But will add some more tests to validate that whole arithmetic expression is replaced by outer ref & not just a subexpr.